rp2/machine_spi: Detect RX overrun when using DMA transfers.

sfe-SparkFro opened the pull request:
### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->
Fixes #18471

### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->
Before this change, the reproduction code in #18471 freezes within ~100 loop iterations. After this change, ~the reproduction code does not freeze (at least, not within 10k loop iterations)~ an `OSError` is raised instead of freezing.

### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

     Delete this heading if not relevant (i.e. small fixes) -->
~I don't know the original motivation for using DMA channels for large SPI transfers, but one downside of this change is that long interrupts could hold up SPI transfers. I don't know how problematic that really is, but users should be writing their interrupts to be efficient anyways.~

~If a user really want to use DMA channels for large SPI transfers, they can easily do this with the existing MicroPython API. Something like this should work (I've not tested it myself):~

```
src_data = bytearray(1024)
DATA_REQUEST_INDEX = 24

spi = machine.SPI(0)

dma = rp2.DMA()
ctrl = dma.pack_ctrl(size=0, inc_write=False, treq_sel=DATA_REQUEST_INDEX)

dma.configure(
    read=src_data,
    write=spi,
    count=len(src_data),
    ctrl=ctrl,
    trigger=True
)
```

An alternative solution could be to enable to the user to select whether they want to use the DMA for large transfers, either when the SPI constructor is called, or when the read/write functions are called.

dpgeorge commented on ports/rp2/machine_spi.c:
```diff
@@ -261,61 +261,14 @@ static void machine_spi_init(mp_obj_base_t *self_in, size_t n_args, const mp_obj
-    const size_t dma_min_size_threshold = 32;
-    int chan_tx = -1;
-    int chan_rx = -1;
-    if (len >= dma_min_size_threshold) {
```
Considering the bug can only (so far?) be reproduced on boards with PSRAM, maybe the best approach is to change the logic here so that it never uses DMA if `src` or `dest` point to PSRAM?

dpgeorge replied:
OK, I see.

I'll have to think about this more, but in general I think it's worth keeping the SPI DMA code.  So how about making the logic be "don't use DMA if on an RP2350 with PSRAM"?

sfe-SparkFro replied:
Fair enough, I can probably wrap the DMA code in an #ifdef with the relevant PSRAM macro (not at my computer right now).

Just curious, what's the motivation for using DMA channels for SPI transfers?

sfe-SparkFro committed:
rp2/machine_spi: Detect RX overrun when using DMA transfers.

Fixes bug where SPI can freeze if another DMA channel is
tranferring to/from PSRAM.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

sfe-SparkFro committed:
rp2/machine_spi: Improve RX overrun handling.

Need to clear the overrun flag.
RX overruns are not a problem if only writing.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

sfe-SparkFro commented:
> It would be instructive if you could test this on a board with PSRAM to confirm that it can be used to fix the problem

Thanks! Just tested it out. Pacing the PSRAM DMA at 20 million transfers per second does not help (my test code typically freezes on the first loop iteration now that I'm transferring from PSRAM to SRAM). If I drop it to 2 million, then it does help, but does not fix the problem (first run froze after ~600 loop iterations).

Maybe it can be fixed by further reducing the transfer rate, but that would really start throttling the DMA, and I'm not even sure that pacing the PSRAM DMA transfer can be guaranteed to fix this. When the DMA encounters a bus stall during a PSRAM transfer, the amount of time its stalled is variable. I'm guessing it's possible for a single very long bus stall to cause an RX FIFO overrun, in which case, no amount of DMA pacing fixes the problem. Makes it a lot less likely, but not guaranteed. Or at least, I'm guessing that's the case, but even if not, you're sacrificing DMA throughput. Which is why I'd like an option to force software SPI transfers, because it's guaranteed to not have overruns while allowing the PSRAM DMA channel to transfer as fast as the XIP subsystem allows.

dpgeorge commented:
> Which is why I'd like an option to force software SPI transfers

You could always use `machine.SoftSPI`, although that's pretty limited in throughput due to the bit banging nature of it.

So, yes, I think it makes sense for a follow-up PR to add a `dma` option to the `machine.SPI` constructor on the rp2 port (stm32's `pyb.I2C` class actually has such a parameter for the same thing).

dpgeorge reviewed (approved):
This change looks reasonable now.  Users will be surprised if the SPI read raises an exception, but that is better than locking up.

nickovs commented:
@dpgeorge While you're at it, can you take a look at the above-mentioned #18622 since it is generally useful even when not trying to pace transfers to PSRAM.

sfe-SparkFro commented:
> So, yes, I think it makes sense for a follow-up PR to add a `dma` option to the `machine.SPI` constructor on the rp2 port (stm32's `pyb.I2C` class actually has such a parameter for the same thing).

Tracking in https://github.com/micropython/micropython/issues/19186

dpgeorge commented:
Rebased, squashed and merged in 6fa7d5be270fa4133f10112c1ed3b501ecbf69d2
